File: examples/android/src/main/java/dev/bluefalcon/viewModels/DeviceViewModel.kt
package dev.bluefalcon.viewModels

import android.util.Log
import dev.bluefalcon.BlueFalconApplication
import dev.bluefalcon.BluetoothPeripheral
import dev.bluefalcon.BluetoothService
import dev.bluefalcon.adapters.DeviceAdapter
import dev.bluefalcon.observables.StandardObservableProperty
import dev.bluefalcon.views.DeviceActivityUI
import dev.bluefalcon.activities.DeviceActivity
import dev.bluefalcon.services.BluetoothServiceConnectedDeviceDelegate

class DeviceViewModel(
    private val deviceActivity: DeviceActivity,
    var bluetoothPeripheral: BluetoothPeripheral
) : BluetoothServiceConnectedDeviceDelegate {

    val deviceActivityUI = DeviceActivityUI(this)
    val services: List<BluetoothService> get() = bluetoothPeripheral.services
    val deviceAdapter = DeviceAdapter(this)
    var connectionStatus = StandardObservableProperty("Connecting...")

    val title: String
        get() = bluetoothPeripheral.bluetoothDevice.address + " " +
                if (bluetoothPeripheral.bluetoothDevice.name != null) bluetoothPeripheral.bluetoothDevice.name else ""

    init {
        BlueFalconApplication.instance.bluetoothService.connect(bluetoothPeripheral)
    }

    fun destroy() {
        BlueFalconApplication.instance.bluetoothService.disconnect(bluetoothPeripheral)
        removeDelegate()
    }

    fun addDelegate() {
        BlueFalconApplication.instance.bluetoothService.connectedDeviceDelegates[bluetoothPeripheral.bluetoothDevice.address] = this
    }

    fun removeDelegate() {
        BlueFalconApplication.instance.bluetoothService.connectedDeviceDelegates.remove(
            bluetoothPeripheral.bluetoothDevice.address
        )
    }

    override fun connectedDevice() {
        connectionStatus.value = "Connected"
    }

    override fun discoveredServices(bluetoothPeripheral: BluetoothPeripheral) {
        Log.v("Bluefalcon", "Services in view model ${bluetoothPeripheral.services}")
        this.bluetoothPeripheral = bluetoothPeripheral
        deviceActivity.runOnUiThread {
            deviceAdapter.notifyDataSetInvalidated()
            deviceAdapter.notifyDataSetChanged()
        }
    }

}

File: examples/MacOS-Example/MacOS-Example/ViewModels/DeviceViewModel.swift
//
//  DeviceViewModel.swift
//  Blue-Falcon
//
//  Created by Andrew Reed on 24/08/2019.
//  Copyright © 2019 Andrew Reed. All rights reserved.
//

import Foundation
import BlueFalcon
import CoreBluetooth
import Combine

class DeviceViewModel: ObservableObject {

    @Published var state: CBPeripheralState = .connecting
    @Published var deviceServiceCellViewModels: [DeviceServiceCellViewModel] = []

    let device: BluetoothPeripheral
    let title: String

    init(device: BluetoothPeripheral) {
        self.device = device
        title = "\(device.bluetoothDevice.identifier.uuidString) \(device.name ?? "")"
    }

    func onAppear() {
        AppDelegate.instance.bluetoothService.connectedDeviceDelegates.append((device.bluetoothDevice.identifier, self))
        connect()
    }

    func onDisapear() {
        AppDelegate.instance.bluetoothService.removeConnectedDeviceDelegate(delegate: self)
    }

    func connect() {
        AppDelegate.instance.bluetoothService.connect(bluetoothPeripheral: device)
    }

}

extension DeviceViewModel: BluetoothServiceConnectedDeviceDelegate {

    func connectedDevice() {
        state = .connected
    }

    func discoveredServices() {
        guard let services = device.bluetoothDevice.services else { return }
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
            self.deviceServiceCellViewModels = services.map({ service -> DeviceServiceCellViewModel in
                DeviceServiceCellViewModel(id: service.uuid, service: service)
            })
        }
    }

}
